source/handbook/cpp-coding-guidelines.rst

```diff
@@ -1,289 +1,238 @@
+7. **ALWAYS** pass non-trivial parameters by reference.
+8. **PREFER** a composite return value (e.g. ``pair``, POD ``struct``) to multiple out parameters.
+9. **PREFER** ``auto`` whenever possible, but use as many type modifiers (``const``, ``volatile``, ``&``, ``*``) as possible.
+10. **ALWAYS** use ``using`` statements instead of ``typedef``
+11. **USE** function qualifiers (``override``, ``final``, ``const``) liberally.
+12. **PREFER** an enumeration to multiple booleans.
+13. **USE** anonymous namespaces liberally in cpp files and export the minimum number of names from each compilation object.
+14. **NEVER** use static free functions.
+
+Numeric Types
+-------------
 
-   What's needed in ``.cpp`` file should only be included in ``.cpp``.
+1. **USE** ``size_t`` whenever dealing with data sizes.
+2. **USE** types defined in **stdint.h** (``uint8_t``, ``uint16_t``, ``uint32_t``, etc.)
+3. **AVOID** using signed types unless it's really necessary and reasonable.
+4. **AVOID** using floating point arithmetic, especially in any part of consensus.
 
-3. Order of includes (top-down)
+File / Directory Layout
+***********************
 
-   1. OWN (local ones)
+If a cpp/h file contains a single class, the file should be named after the class and be UpperCamelCase.
+If a cpp/h file contains utility functions, the file should be given a descriptive name and be UpperCamelCase.
 
-   2. Project common
+We expect parallel source and test directory hierarchies.
+For example,
 
-   3. Shared/common ``<core/...>``
+::
 
-   4. System/STL
+    root
+    - parser  # package name, directory containing source code
+      Parser.h/cpp  # contains class Parser
+      - ast
+        - Ast.h/cpp  # contains various exports
 
-Nice link for further reading: http://www.topology.org/linux/include.html
+    - tests  # directory containing tests for source code
+      - ParserTests.cpp  # unit tests for parser/Parser.h/cpp
+      - ast
+        - AstTests.cpp  # unit tests for parser/ast/Ast.h/cpp
 
-if/for
+Naming
 ******
 
-* Do not use such a construct when ``for`` doesn't have a body
-
-  .. code-block:: c++
+1. Static and free functions should be UpperCamelCase
 
-     for (a; b; c);
+2. Member function names should be lowerCamelCase
 
-  Instead use
+3. Constants (global or local) should be Upper_Camel_Case.
 
-  .. code-block:: c++
+4. #defines should be TITLE_CASE
 
-     for (a; b; c) {}
+5. Local variables and function parameters should be lowerCamelCase
 
-This leaves clear intention of what you had in mind.
+6. Struct members should be UpperCamelCase
 
-Operators
-*********
 
-* In case of operators please put additional space before and after them. This makes code much more readable.
+Prefixes
+--------
 
-  This should be always used in case of ``=``, ``==``, ``!=``, ``&&``, ``||``. So this one's ok:
+1. Prefix class instance members with ``m_``
 
-  .. code-block:: c++
+2. Prefix class static members with ``s_``
```

Comment:
hmm, I don't remember do we do this anywhere?

Reply:
it's rare, but we have this naming in PacketSocket